[REVIEW] First step for NetworkX compatibility

seunghwak opened the pull request:
First step to address https://github.com/rapidsai/cugraph/issues/192

added

add_edge_from(), add_weighted_edges_from(), from_cudf_edgelist(), clear(), and number_of_nodes()

ChuckHastings commented on python/cugraph/graph/convert_matrix.pyx:
```diff
@@ -0,0 +1,50 @@
+
+def from_cudf_edgelist(df, source='source', target='target', weight=None):
+    """
+    Add edges to a graph. Currently, self should be an empty graph.
```
I'm confused.  The comment suggests that an error will be raised if self is not empty.  The code does nothing with self, creates an empty graph, adds edges and returns the new graph object.  This seems inconsistent.

seunghwak replied:
A great catch, I forgot to update comments after moving this from inside the Graph class (as a member function) to outside the Graph class. I will update the comments.

seunghwak replied:
And I should remove add_edges_from() and add_weighted_edges_from() as discussed in the email, so no need to review those functions (I didn't know this PR will be reviewed at this point, I was waiting for cudf's fix so all the tests can be passed before starting a review).

aschaffer commented on python/cugraph/graph/test_graph.py:
```diff
@@ -75,6 +76,40 @@ def compare_offsets(offset0, offset1):
     return True
 
 
+def compare_graphs(nx_graph, cu_graph):
```
compare_graphs() is not exactly correct. You can have 2 identical graphs, one with vertex indices permuted wrt the other and they may have different row_offsets and column_indices. This is the problem of graph isomorphism and does _not_ currently have (such) a trivial solution (it's active research). Perhaps this function should be renamed and comments added to reflect that is a very simplistic heuristics for graph comparison.

ChuckHastings replied:
To clarify... this compare_graphs will return the correct value for it's intended purpose.  If you take the same set of edges, with the same numbering of vertices and create a graph, this function will return true if the resulting graphs match and false if they don't.

The issue @aschaffer is mentioning is a broader graph comparison issue.

seunghwak replied:
Thank you very much for the comments, and yes, this is not to test two graphs are isomorphic but to test two graphs are automorphic (isomorphic + vertex mapping between two graphs is an identity mapping). I will add comments to make this point clear.

seunghwak commented:
Why's this PR in the Waiting on Author state? 

python/cugraph/graph/test_graph.py

is not marked as outdated may be due to my force-push (I should do that because I accidentally added hornet integration related updates to this branch) but I addressed the comments in more recent commits.

aschaffer commented:
Is there any action for me here? I thought I've already approved this.

afender commented:
> Is there any action item for me here? I thought I've already approved this.

It is best to select "approve" when you approve the PR so that others can see that you approved. In this case, your status is on "commented" for instance

aschaffer commented:
The intention was to comment _without_ blocking the merge (i.e., the comment is merely a suggestion and not a reason to prevent the merge). The "Comment" option, in Review, doesn't prevent merging, does it?
